doc/config-generic-update-rollback/SONiC_Generic_Config_Update_and_Rollback_Design.md

```diff
@@ -0,0 +1,867 @@
+|side-effects|None                     |
+|assumptions |running-config locked    | The implementor of this contract might interact with ConfigDB to get the running-config, it is assumed the ConfigDB is locked for changes for the lifespan of the operation.
+
+If `order-patch` has to force the update to follow very specific steps, it would have to provide multiple JsonChange objects in the return list of `order-patch`.
+
+`order-patch` is returning a list of JsonChanges instead of a simple JsonPatch with multiple operations because a JsonChange can group together multiple JsonPatch operations that share no dependency and can be executed together. This can help the implementor of `apply-change` to optimize the mechanism for applying JsonChange e.g. group changes under same parent together or reduce number of service restarts.
+
+For example:
+Assume JsonPatch contains:
+```
+[
+  { "op": "add", "path": "/DHCP_SERVERS/4.4.4.4", "value": {} },
+  { "op": "replace", "path": "/DEVICE_NEIGHBOR/Ethernet8/port", "value": "eth1" },
+  { "op": "add", "path": "/DHCP_SERVERS/2.2.2.2", "value": {} }
+]
+```
+We have 2 operations updating DHCP servers, and another operation for DEVICE_NEIGHBOR. We can assume DHCP and DEVICE_NEIGHBOR tables to be independent meaning they can updated at the same time. `order-patch` would return:
+```
+[
+  [
+    { "op": "add", "path": "/DHCP_SERVERS/4.4.4.4", "value": {} },
+    { "op": "replace", "path": "/DEVICE_NEIGHBOR/Ethernet8/port", "value": "eth1" },
+    { "op": "add", "path": "/DHCP_SERVERS/2.2.2.2", "value": {} }
+  ],
+]
+```
+Updating DHCP_SERVERS requires restarting `dhcp_relay` service, so if the above patch is to be executed in order, we will restart `dhcp_relay` service twice.
+But since the implementor of `apply-change` can order the operations in any way they see fit since they are OK to update together. They can decide move the DHCP updates together, and DHCP table twice, but restart `dhcp_relay` service only once.
+
+Let's take a visual example, assume we have a JsonPatch with 8 operations, and here is the topological order of the operation. Arrow from op-x to op-y means op-y depends on op-x.
+
+<img src="files/jsonpatch-topology.png" alt="jsonpatch-topology" width="800"/>
+
+If we sort by on the operations we would have:
+
+<img src="files/jsonpatch-operations-order.png" alt="jsonpatch-operations-order" width="800"/>
+
+But if we organize the operations into groups of JsonChange, we will have:
+
+<img src="files/jsonchanges-order.png" alt="jsonchanges-order" width="800"/>
+
+This will allow the the implementor of `apply-change` to have the freedom to optimize the operations in any order they see fit.
+
+**NOTE:** Check Patch Orderer implementation design design details in [Json_Patch_Ordering_using_YANG_Models_Design](Json_Patch_Ordering_using_YANG_Models_Design.md) document.
+
```

Comment:
Even though we can control the order of operation from outside, I feel, until and unless, backend handles the dependency we can still result in a service crash.

We have noticed some circumstances where the previous config changes are not processed by the backend and  new config changes come which are already in the pipeline and being processed before last config change.

Reply:
We have isolated this concern behind its own contract called `Apply-Change`. The implementation of this contract should take care of applying the changes to ConfigDB, restart corresponding services, and making sure services have absorbed the change correctly. Check the design doc: Json_Change_Application_Design.md for details.